LDAP error Unable to extract user primary group with gidNumber attribute from user entries with ActiveDirectory

I went ahead and stood up a Rocky 9 server for the gateway and I'm still having issues with authentication against Active Directory. For the sake of troubleshooting, I statically set sAMAccountName in place of uid in ldap.py. This got me a new error: `[ERROR] ⸬slurmweb.apps.ldap:45  ↦ LDAP error: Unable to extract user primary group with gidNumber attribute from user entries` 

Navigating to the login page and entering credentials, I get:

<img width="523" alt="Screenshot 2024-08-23 at 1 23 57 PM" src="https://github.com/user-attachments/assets/4c524be9-1237-4f53-8e70-0c9fd5ab20be">
<img width="286" alt="Screenshot 2024-08-23 at 1 24 09 PM" src="https://github.com/user-attachments/assets/f3a893e4-0d60-473b-a7a1-291b9e8b474f">

Thanks!

_Originally posted by @rseaman2016 in https://github.com/rackslab/Slurm-web/issues/340#issuecomment-2307605432_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - pkgs: Add requirement on RFL.core >= 1.0.3.
 - frontend: Add intermediate cluster list width to 80% on large screens, before
   going down to 60% on even larger screens.
+- docs: Update configuration reference documentation.
 
 ### Fixed
 - backend: handle rfl.settings.errors.SettingsSiteLoaderError when loading

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Explain `[slurmrestd]` > `version` agent setting is more intended for
     developers and should not be changed.
   - Mention Slurm accounting is required in quickstart guide (#341).
+- conf: Add `ldap`>`user_primary_group_attribute` for the gateway.
+- gateway: Support custom LDAP user primary group attribute.
+- ldap-check: Support custom LDAP user primary group attribute.
 
 ## Changed
 - pkgs: Add requirement on RFL.core >= 1.0.3.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,9 +25,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   classes (#342).
 
 ## Changed
-- pkgs: Add requirement on RFL.core >= 1.0.3.
 - frontend: Add intermediate cluster list width to 80% on large screens, before
   going down to 60% on even larger screens.
+- pkgs: Add requirement on RFL.core and RFL.authentication >= 1.0.3.
 - docs: Update configuration reference documentation.
 
 ### Fixed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,9 +16,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Explain `[slurmrestd]` > `version` agent setting is more intended for
     developers and should not be changed.
   - Mention Slurm accounting is required in quickstart guide (#341).
-- conf: Add `ldap`>`user_primary_group_attribute` for the gateway.
-- gateway: Support custom LDAP user primary group attribute.
-- ldap-check: Support custom LDAP user primary group attribute.
+- conf:
+  - Add `ldap`>`user_primary_group_attribute` parameter for the gateway.
+  - Add `ldap`>`group_object_classes` parameter for the gateway.
+- gateway: Support custom LDAP user primary group attribute and group object
+  classes (#342).
+- ldap-check: Support custom LDAP user primary group attribute and group object
+  classes (#342).
 
 ## Changed
 - pkgs: Add requirement on RFL.core >= 1.0.3.

--- a/docs/modules/conf/examples/gateway.ini
+++ b/docs/modules/conf/examples/gateway.ini
@@ -150,10 +150,24 @@ user_name_attribute=uid
 # Default value: cn
 user_fullname_attribute=cn
 
+# User entry attribute for primary group ID
+#
+# Default value: gidNumber
+user_primary_group_attribute=gidNumber
+
 # Group entry attribute for name
 #
 # Default value: cn
 group_name_attribute=cn
+
+# List of LDAP object classes for groups
+#
+# Default value:
+# - posixGroup
+# - groupOfNames
+group_object_classes=
+  posixGroup
+  groupOfNames
 
 # DN used to bind to the LDAP server. When this parameter is not defined,
 # access to LDAP directory is performed anonymously.

--- a/docs/modules/conf/partials/conf-gateway.adoc
+++ b/docs/modules/conf/partials/conf-gateway.adoc
@@ -366,6 +366,17 @@ _No default value_
 
 |-
 
+|user_primary_group_attribute
+|str
+|User entry attribute for primary group ID
+
+
+
+
+*Default:* `gidNumber`
+
+|-
+
 |group_name_attribute
 |str
 |Group entry attribute for name
@@ -374,6 +385,23 @@ _No default value_
 
 
 *Default:* `cn`
+
+|-
+
+|group_object_classes
+|list[str]
+|List of LDAP object classes for groups
+
+
+
+
+*Default:*
+
+
+* `posixGroup`
+
+* `groupOfNames`
+
 
 |-
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -16,7 +16,7 @@ dependencies = [
     "Flask",
     "requests",
     "aiohttp",
-    "RFL.authentication >= 1.0.2",
+    "RFL.authentication >= 1.0.3",
     "RFL.core >= 1.0.3",
     "RFL.log",
     "RFL.settings",

--- a/slurmweb/apps/gateway.py
+++ b/slurmweb/apps/gateway.py
@@ -119,6 +119,7 @@ class SlurmwebAppGateway(SlurmwebWebApp, RFLTokenizedWebApp):
                     user_fullname_attribute=self.settings.ldap.user_fullname_attribute,
                     user_primary_group_attribute=self.settings.ldap.user_primary_group_attribute,
                     group_name_attribute=self.settings.ldap.group_name_attribute,
+                    group_object_classes=self.settings.ldap.group_object_classes,
                     cacert=self.settings.ldap.cacert,
                     starttls=self.settings.ldap.starttls,
                     bind_dn=self.settings.ldap.bind_dn,

--- a/slurmweb/apps/gateway.py
+++ b/slurmweb/apps/gateway.py
@@ -117,6 +117,7 @@ class SlurmwebAppGateway(SlurmwebWebApp, RFLTokenizedWebApp):
                     user_class=self.settings.ldap.user_class,
                     user_name_attribute=self.settings.ldap.user_name_attribute,
                     user_fullname_attribute=self.settings.ldap.user_fullname_attribute,
+                    user_primary_group_attribute=self.settings.ldap.user_primary_group_attribute,
                     group_name_attribute=self.settings.ldap.group_name_attribute,
                     cacert=self.settings.ldap.cacert,
                     starttls=self.settings.ldap.starttls,

--- a/slurmweb/apps/ldap.py
+++ b/slurmweb/apps/ldap.py
@@ -30,6 +30,7 @@ class SlurmwebAppLDAPCheck(SlurmwebGenericApp):
             user_fullname_attribute=self.settings.ldap.user_fullname_attribute,
             user_primary_group_attribute=self.settings.ldap.user_primary_group_attribute,
             group_name_attribute=self.settings.ldap.group_name_attribute,
+            group_object_classes=self.settings.ldap.group_object_classes,
             starttls=self.settings.ldap.starttls,
             bind_dn=self.settings.ldap.bind_dn,
             bind_password=self.settings.ldap.bind_password,

--- a/slurmweb/apps/ldap.py
+++ b/slurmweb/apps/ldap.py
@@ -28,6 +28,7 @@ class SlurmwebAppLDAPCheck(SlurmwebGenericApp):
             group_base=self.settings.ldap.group_base,
             user_name_attribute=self.settings.ldap.user_name_attribute,
             user_fullname_attribute=self.settings.ldap.user_fullname_attribute,
+            user_primary_group_attribute=self.settings.ldap.user_primary_group_attribute,
             group_name_attribute=self.settings.ldap.group_name_attribute,
             starttls=self.settings.ldap.starttls,
             bind_dn=self.settings.ldap.bind_dn,

